Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -61,69 +62,48 @@ final class ConnectionPool {
     ///
     /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
-    /// is used to determine if there already exists an associated `HTTP1ConnectionProvider` in `connectionProviders`.
+    /// is used to determine if there already exists an associated `HTTP1ConnectionProvider` in `providers`.
     /// If there is, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
-    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?, setupComplete: EventLoopFuture<Void>) -> EventLoopFuture<Connection> {
         let key = Key(request)
 
-        let provider: HTTP1ConnectionProvider = self.connectionProvidersLock.withLock {
-            if let existing = self.connectionProviders[key] {
-                existing.stateLock.withLock {
-                    existing.state.pending += 1
-                }
+        let provider: HTTP1ConnectionProvider = self.lock.withLock {
+            if let existing = self.providers[key], existing.enqueue() {
                 return existing
             } else {
-                let http1Provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, parentPool: self)
-                self.connectionProviders[key] = http1Provider
-                http1Provider.stateLock.withLock {
-                    http1Provider.state.pending += 1
-                }
-                return http1Provider
+                let provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, pool: self)
+                self.providers[key] = provider
+                return provider
             }
         }
 
-        return provider.getConnection(preference: preference)
+        return provider.getConnection(preference: preference, setupComplete: setupComplete)
     }
 
-    func release(_ connection: Connection) {
-        let connectionProvider = self.connectionProvidersLock.withLock {
-            self.connectionProviders[connection.key]
-        }
-        if let connectionProvider = connectionProvider {
-            connectionProvider.release(connection: connection)
+    func delete(_ provider: HTTP1ConnectionProvider) {
+        self.lock.withLockVoid {
+            self.providers[provider.key] = nil
         }
     }
 
-    func prepareForClose(on eventLoop: EventLoop) -> EventLoopFuture<Void> {
-        let connectionProviders = self.connectionProvidersLock.withLock {
-            self.connectionProviders.values
+    func close(on eventLoop: EventLoop) -> EventLoopFuture<Bool> {
+        let providers = self.lock.withLock {
```

Comment:
shouldn't `self.providers` be emptied here? Ideally we'd transition to a state that doesn't have providers but that's maybe for a followup too?

Reply:
all providers will delete themselves (and we assert that in client's `deinit`)

Reply:
@artemredkin ok, what prevents a second call to `close(on:)`?

Reply:
Also, wouldn't it be better to just remove them here? What's the benefit of them deleting themselves?

Reply:
second call to close is prevented by the client, and since providers already delete themselves when they are empty (which is basically what close means, we close/fail all connections waiters and suddenly provider is empty so it deletes itself)